### PR TITLE
fortran/use-mpi-f08: add support for Fortran 2018 ISO_Fortran_binding.h

### PR DESCRIPTION
**Pull request by ggouaillardet**
add the infrastructuce and update MPI_Send() and MPI_Recv()

Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>

### CODE REVIEW

**Comment by ggouaillardet** [comment]
@jsquyres @hppritcha @bosilca this is a first cut on a topic we discussed long time ago.

stuff in `ompi/mpi/fortran/use-mpi-f08/cdesc` is (inspired) from MPICH

I ran a simple pingpong test with
```
integer :: buf(0:1025,0:1025)
```
and use `buf(1:1024,1:1024)` as the buffer. The performance is now 4x faster with `use mpi_f08` compared to `use mpi` (that requires internal buffer allocation/filling/deallocation from the Fortran runtime). I used Intel compilers since upcoming gcc 9 is not quite ready now (and this is currently discussed on the MP starting at https://gcc.gnu.org/ml/fortran/2019-04/msg00013.html)

**Comment by bosilca** [comment]
This is really cool stuff, it was about time someone address this issue. Let me take a look at how we can build these types in a better way. Can you point me to some good documentation about the Fortran cdesc ?

**Comment by ggouaillardet** [comment]
@bosilca I guess the best documentation is in the standard. I am not sure it is available at no cost, but you can check the recent and freely available draft at https://j3-fortran.org/doc/year/18/18-007r1.pdf

As a side note, we should really fix how non blocking collectives handle datatypes/operators refcounts
(if I apply the same technique to non blocking collectives, when we will end up using free'd datatypes and bad things will happen).

**Comment by ggouaillardet** [comment]
Refs. open-mpi/ompi#2154

File: ompi/mpi/fortran/use-mpi-f08/cdesc/send_cdesc.c [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+{
+    int c_ierr;
+
+    MPI_Comm c_comm = PMPI_Comm_f2c(*comm);
+    MPI_Datatype c_datatype, c_type = PMPI_Type_f2c(*datatype);
+    void *buf = x->base_addr;
+    int c_count =  OMPI_FINT_2_INT(*count);
+
+    c_datatype = c_type;
+    if (x->rank != 0 && !CFI_is_contiguous(x)) {
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/cdesc/send_cdesc.c]: What is `CFI_is_contiguous`?  I can't find it anywhere in our code base.  Is it a Fortran compiler-provided function?

File: ompi/mpi/fortran/use-mpi-f08/cdesc/cdesc.c [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        if (cdesc->dim[i].sm == accum_sm) {
+            mpi_errno = PMPI_Type_contiguous(extent, types[i], &types[i+1]);
+        } else {
+            mpi_errno = PMPI_Type_create_hvector(extent, 1, cdesc->dim[i].sm, types[i], &types[i+1]);
+        }
+        if (mpi_errno != MPI_SUCCESS) {
+            last = i; goto fn_exit;
+        }
+
+        mpi_errno = PMPI_Type_commit(&types[i+1]);
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/cdesc/cdesc.c]: Don't we only have to commit the *last* datatype?

I.e., isn't "commit" potentially an expensive operation?  If so, can't we use uncommitted datatypes to build up new datatypes, and therefore only have to commit the *last* datatype?

File: ompi/mpi/fortran/use-mpi-f08/cdesc/send_cdesc.c (hunk 1 above)
Comment (bosilca) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/send_cdesc.c]: [Quite interesting...](https://software.intel.com/en-us/fortran-compiler-developer-guide-and-reference-cfi-is-contiguous)

Comment (jsquyres) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/send_cdesc.c]: Just to be triple sure, we might want to also add a `configure` test for `CFI_is_contiguous`...?  (I know it should probably be there if `ISO_Fortran_binding.h` is there, but it wouldn't hurt to check (because we've seen compile bugs before!).

File: ompi/mpi/fortran/use-mpi-f08/cdesc/cdesc.c (hunk 2 above)
Comment (jsquyres) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/cdesc.c]: @bosilca confirmed to me in an email that I was correct -- you only need to commit the final datatype.  Depending on the complexity of the array slice, this could be a nice optimization.

File: ompi/mpi/fortran/use-mpi-f08/cdesc/send_cdesc.c (hunk 1 above)
Comment (ggouaillardet) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/send_cdesc.c]: `CFI_is_contiguous()` and the `CFI_cdesc_t` type are defined in `ISO_Fortran_binding.h`. I will have to enhance the test and also run a test program. I do not think the standard specify an ABI, so we should detect some mismatch such as GNU C and Intel Fortran compilers.

File: ompi/mpi/fortran/use-mpi-f08/cdesc/cdesc.c (hunk 2 above)
Comment (ggouaillardet) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/cdesc.c]: yep, I will do that (and issue a PR to mpich since this is basically their code).

**Comment by bosilca** [comment]
I did not meant to resolve the conversation, I just clicked on the wrong button. What I wanted to say is that I do not like copying the code from MPICH here. This entire piece of code seems to be a call to darray, we might want to go that route.

File: ompi/mpi/fortran/use-mpi-f08/cdesc/send_cdesc.c (hunk 1 above)
Comment (jsquyres) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/send_cdesc.c]: I don't know if you need to run a test program -- if we can `AC_LINK_IFELSE` (or even just `AC_CHECK_FUNC`), that should probably be enough, right?

Comment (ggouaillardet) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/send_cdesc.c]: I do not think the `CFI_cdesc_t` structure and other subroutine are ABI compatibles between compilers, so I believe some more tests are needed. GNU & Fortran compilers do not interoperate (Intel compilers `#define CFI_is_contiguous(dv) for_CFI_is_contiguous(dv)`) so linking C and Fortran binary is enough to detect the incompatibility). An other combination of compilers might link but crash at runtime, so I am afraid that at the end, we have to build __and__ run a test program.

Comment (jsquyres) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/send_cdesc.c]: Hmm.  I hesitate to `AC_RUN_IFELSE` because of the difficulties that raises with cross-compilation.  If we can avoid `AC_RUN_IFELSE`, we should.

**Comment by AboorvaDevarajan** [comment]
Can one of the admins verify this patch?

**Comment by ggouaillardet** [comment]
These new Fortran 2008 bindings bring new opportunities (performance improvements, correct handling of sub-arrays in non blocking subroutines) but also some challenges I will try to list here 

 1. handling a part of a subarray
```fortran
integer :: a(0:3,0:3)
call mpi_send(a(1:2,1:2), 3, MPI_INT, ...)
```
  this is currently unsupported. Under the hood, a derived datatype is created for the subarray (4 elements) and the `C` `PMPI_Send()` binding is invoked with this intermediary datatype and `count=1`. As a consequence, only the full subarray can be sent.
 2. scatter and friends
```fortran
integer :: a(0:3,0:3)
call mpi_scatter(a(1:2,1:2), 2, MPI_INT, ...)
```
if the communicator size is `2`, then this is a legit usage that is currently not supported (just as above, the `C` binding is invoked with a ddt for the full subarray and `count=1`. In order to work, the intermediate datatype should have 2 elements (and not 4).
if the communicator size is `4`, this is again a legit call, but would have to invoke the `C` `PMPI_Scatterv()` binding which is currently not supported.
 3. pack and friends
```fortran
integer :: a(0:3,0:3)
call mpi_pack(..., outbuf=a(0:2,0:2), outsize=4, ...)
```
this is arguably a dumb thing to do, and this is currently not supported. note the current `ignore_tkr` version works just fine.
 4. `ignore_tkr` support
Currently, `ignore_tkr` is required to build the `mpi_f08` bindings. A compiler might not support `ignore_tkr` but might support the new bindings (and yes, I have one in mind).

Bottom line, some decisions have to be made about what we support and how.
For example, we might not handle (non contiguous) subarrays and have the subroutine fail with `MPI_ERR_INTERN` or `MPI_ERR_BUFFER`.
About the scatter example, we might not support this (and return with the same error as above), fully support it or simply fall back to the old bindings (but `MPI_Iscatter()` might be still broken).
About the pack example, we might use the old bindings or simply not support that (that is a dumb thing to do anyway).
Of course, if the compiler does not support the old `ignore_tkr` bindings, we do not have the option to fall back on these bindings.

As a reminder, I opened pending #2154 to correctly handle derived datatypes freed in the middle of a collective operation (can be done at the component level e.g. `coll/libnbc` or at the MPI level once for all components). Without this PR, non blocking collectives would either be busted or leak a datatype.

@jsquyres @bosilca @kawashima-fj could you please comment on that ?
I hope we can reach a consensus on how to move forward (e.g. how and with which limitations).

**Comment by jsquyres** [comment]
bot:lanl:retest

File: ompi/mpi/fortran/use-mpi-f08/cdesc/cdesc.c [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "cdesc.h"
+
+int cdesc_create_datatype(CFI_cdesc_t *cdesc, int oldcount, MPI_Datatype oldtype, MPI_Datatype *newtype)
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/cdesc/cdesc.c]: Should this be `ompi_cdest_create_datatype()` (i.e., `ompi_` prefix)?

File: ompi/mpi/fortran/use-mpi-f08/cdesc/cdesc.c (hunk 2 above)
Comment (jsquyres) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/cdesc.c]: Still need to move this `type_commit` outside of the loop and only commit the last/final datatype.

File: ompi/mpi/fortran/use-mpi-f08/cdesc/cdesc.c [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            done = 1;
+        }
+
+        if (cdesc->dim[i].sm == accum_sm) {
+            mpi_errno = PMPI_Type_contiguous(extent, types[i], &types[i+1]);
+        } else {
+            mpi_errno = PMPI_Type_create_hvector(extent, 1, cdesc->dim[i].sm, types[i], &types[i+1]);
+        }
+        if (mpi_errno != MPI_SUCCESS) {
+            last = i; goto fn_exit;
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/cdesc/cdesc.c]: Nit: please don't have 2 statements on a single line (same comment above where the same pattern occurs).

File: ompi/mpi/fortran/use-mpi-f08/cdesc/Makefile.am [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+noinst_LTLIBRARIES = $(module_sentinel_file)
+
+# f08 support modules
+
+libforce_usempif08_internal_cdesc_to_be_built_la_SOURCES = \
+        bindings.h \
+        cdesc.h \
+        cdesc.c \
+        \
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/cdesc/Makefile.am]: Nit: remove the extra blank lines.  `cdesc.c` should be the last entry in the list, and not have a `\` continuation character.

File: ompi/mpi/fortran/use-mpi-f08/Makefile.am [hunk 6]
```diff
@@ -44,6 +45,11 @@ module_sentinel_file = \
         mod/libforce_usempif08_internal_modules_to_be_built.la \
         bindings/libforce_usempif08_internal_bindings_to_be_built.la
 
+if OMPI_FORTRAN_HAVE_C_ISO_FORTRAN
+module_sentinel_file += \
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/Makefile.am]: I know I'm being picky here, but it confused me for a while when reading this: the `ibforce_usempif08_internal_cdesc_to_be_built.la` library is *not* a sentinel that is used to force the building of a fortran module: it is the actual library itself.  Adding it to `module_sentinel_file` and naming it `libforce_...blahblahbla` is confusing.

Also, don't you need to slurp of `libforce_ibforce_usempif08_internal_cdesc_to_be_built.la` into `libompi_usmpif08.la` (i.e., as an LT convenience library)?  Otherwise, the `[ompi_]cdesc_create_datatype` symbol won't be available to the greater OMPI code base.

File: ompi/mpi/fortran/configure-fortran-output.h.in [hunk 7]
```diff
@@ -46,6 +46,11 @@
 ! Line 2 of the ignore TKR syntax
 #define OMPI_FORTRAN_IGNORE_TKR_TYPE @OMPI_FORTRAN_IGNORE_TKR_TYPE@
 
+#define OMPI_F08_IGNORE_TKR_TYPE @OMPI_F08_IGNORE_TKR_TYPE@
+
+#define OMPI_F08_BINDINGS_EXTENSION @OMPI_F08_BINDINGS_EXTENSION@
+
+#define OMPI_F08_GCC_ATTRIBUTES(buf) @OMPI_F08_GCC_ATTRIBUTES@
```
Comment (jsquyres) [inline ompi/mpi/fortran/configure-fortran-output.h.in]: I see these 3 macros defined in the `.m4`, above, but what are they used for in the source code?

File: config/ompi_fortran_check_bind_fortran.m4 [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+   write (*,*) is_contiguous(a1)
+   write (*,*) is_contiguous(a2)
+   write (*,*) is_contiguous(a3)
+end program
+EOF
+               OPAL_LOG_COMMAND([$FC $FCFLAGS $FCFLAGS_f90 -o conftest conftest.f90 conftest_c.o $LDFLAGS $LIBS],
+                   [AS_VAR_SET(iso_fortran_binding_var, yes)],
+                   [AS_VAR_SET(iso_fortran_binding_var, no)])],
+              [AS_VAR_SET(iso_fortran_binding_var, no)])
+          cd ..; rm -rf conftest.$$])],
```
Comment (jsquyres) [inline config/ompi_fortran_check_bind_fortran.m4]: Nit: please use 2 lines, not 1.

**Comment by ggouaillardet** [comment]
@jsquyres I addressed all your comments and updated the PR.

File: ompi/mpi/fortran/use-mpi-f08/cdesc/pack_external_cdesc.c [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        c_ierr = ompi_cdesc_create_datatype(x, c_incount, c_type, &c_datatype);
+        if (MPI_SUCCESS != c_ierr) {
+            if (NULL != ierr) *ierr = OMPI_INT_2_FINT(c_ierr);
+            return;
+        }
+        c_incount = 1;
+    }
+    c_ierr = PMPI_Pack_external(c_datarep, OMPI_F2C_BOTTOM(inbuf),
+                               c_incount,
+                               c_type, outbuf,
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/cdesc/pack_external_cdesc.c]: Should `c_type` be `c_datatype`?

File: ompi/mpi/fortran/use-mpi-f08/cdesc/pack_cdesc.c [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+ */
+
+#include "ompi_config.h"
+
+#include "ompi/mpi/fortran/mpif-h/bindings.h"
+#include "ompi/mpi/fortran/use-mpi-f08/cdesc/bindings.h"
+#include "ompi/mpi/fortran/base/constants.h"
+
+void ompi_pack_cdesc(CFI_cdesc_t* x, MPI_Fint *incount, MPI_Fint *datatype,
+		     char *outbuf, MPI_Fint *outsize, MPI_Fint *position,
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/cdesc/pack_cdesc.c]: Can't both the inbuf and outbuf of all these pack/unpack routines be descriptors?

I.e., don't we have to combine the inbuf, the MPI datatype, and the outbuf to do the underlying memcpy(ies) correctly?

File: ompi/mpi/fortran/use-mpi-f08/mod/mpi-f08-interfaces.F90 [hunk 11]
```diff
@@ -580,11 +580,12 @@ subroutine MPI_Pack_f08(inbuf,incount,datatype,outbuf,outsize,position,comm,ierr
    use :: mpi_f08_types, only : MPI_Datatype, MPI_Comm
    implicit none
    !DEC$ ATTRIBUTES NO_ARG_CHECK :: inbuf, outbuf
-   !GCC$ ATTRIBUTES NO_ARG_CHECK :: inbuf, outbuf
+   OMPI_F08_GCC_ATTRIBUTES(inbuf)
+   !GCC$ ATTRIBUTES NO_ARG_CHECK :: outbuf
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/mod/mpi-f08-interfaces.F90]: Per above/comment in the C file, the outbuf can be a descriptor, too.

File: ompi/mpi/fortran/use-mpi-f08/cdesc/Makefile.am [hunk 12]
```diff
@@ -63,7 +63,16 @@ libusempif08_cdesc_la_SOURCES = \
         pack_cdesc.c \
         pack_external_cdesc.c \
         unpack_cdesc.c \
-        unpack_external_cdesc.c
+        unpack_external_cdesc.c \
+        \
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/cdesc/Makefile.am]: Ditto with prior comment: I'd just put the files in an alphabetical list.

**Comment by bosilca** [comment]
Going back to the discussion @ggouaillardet started last week, more specifically on his 3 array shape challenges. On all these points the issue seems to be related to the expected extent of the resulting datatype. Honestly, I am not sure I see what the outcome of ```mpi_send(a(1:2,1:2), 3, MPI_INT, ...)``` is expected to be ? I can see it as adding an extent of 0 to the resulting datatype, and then the count is simply sending the same data multiple times. Or, adding an extent according to one of the dimensions. Both these approaches look legit to me.

**Comment by jsquyres** [comment]
In MPI-3.1 17.1.2 ([see here](https://www.mpi-forum.org/docs/mpi-3.1/mpi31-report/node422.htm#Node422)), it says:

> All nonblocking MPI functions (e.g., MPI_ISEND, MPI_PUT, MPI_FILE_WRITE_ALL_BEGIN) behave as if the user-specified elements of choice buffers are copied to a contiguous scratch buffer in the MPI runtime environment.

BTW, this behavior is only supposed to occur if the compile-time constant `MPI_SUBARRAYS_SUPPORTED` == `.TRUE.`.  I don't recall offhand if that value was set in this PR.  ... @ggouaillardet?

File: ompi/mpi/fortran/use-mpi-f08/cdesc/Makefile.am [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+# This Makefile is only relevant if we're building the "use mpi_f08"
+# MPI bindings.
+if OMPI_BUILD_FORTRAN_USEMPIF08_BINDINGS
+
+AM_CPPFLAGS = -DOMPI_COMPILING_FORTRAN_WRAPPERS=1
+
+###########################################################################
+
+module_sentinel_file =
```
[the remark this answers is not in the document]
Comment (ggouaillardet) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/Makefile.am]: I finally got what you mean(!). Part of the confusion comes from the fact we currently slurp the sentinels into `libmpi_usempif08.la` (!). I just issued https://github.com/open-mpi/ompi/pull/6811 to fix this, can you please review this? I will then fix this PR as you expect.

File: ompi/mpi/fortran/use-mpi-f08/cdesc/bindings.h [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+ *                         and Technology (RIST).  All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#ifndef OMPI_CDESC_BINDINGS_H
+#define OMPI_CDESC_BINDINGS_H
```
[the remark this answers is not in the document]
Comment (ggouaillardet) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/bindings.h]: yes, it is used by **all** the `*_cdesc.c` files.
(I tried to logically group the commits, so some file might only be used by later commits)

File: ompi/mpi/fortran/use-mpi-f08/cdesc/bsend_init_cdesc.c [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+
+    c_datatype = c_type;
+    if (x->rank != 0 && !CFI_is_contiguous(x)) {
+        c_ierr = ompi_cdesc_create_datatype(x, c_count, c_type, &c_datatype);
+        if (MPI_SUCCESS != c_ierr) {
+            if (NULL != ierr) *ierr = OMPI_INT_2_FINT(c_ierr);
+            return;
+        }
+        c_count = 1;
+    }
```
[the remark this answers is not in the document]
Comment (ggouaillardet) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/bsend_init_cdesc.c]: the code contains a `return` statement, so would you be fine using a macro instead?

File: ompi/mpi/fortran/use-mpi-f08/cdesc/pack_cdesc.c (hunk 10 above)
Comment (ggouaillardet) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/pack_cdesc.c]: this is where I asked for guidance in a previous comment.

IMHO, that falls into the category "the enduser wants to do something pretty dumb but that is allowed by the standard".

Anyway, if we make `outbuf` a descriptor, then we cannot simply invoke `PMPI_Pack()` anymore (since it expects a contiguous `outbuf`), so it would be much easier to let the Fortran runtime handle this ... unless the compiler does not support `ignore_tkr`.

If simply returning an error when `outbuf` is not contiguous is not an option **and** we want to support compilers that cannot do `ignore_tkr`, then the only option is to do it by ourself.

File: ompi/mpi/fortran/use-mpi-f08/cdesc/imrecv_cdesc.c [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+    if (x->rank != 0 && !CFI_is_contiguous(x)) {
+        c_ierr = ompi_cdesc_create_datatype(x, c_count, c_type, &c_datatype);
+        if (MPI_SUCCESS != c_ierr) {
+            if (NULL != ierr) *ierr = OMPI_INT_2_FINT(c_ierr);
+            return;
+        }
+        c_count = 1;
+    }
+    c_ierr = OMPI_INT_2_FINT(PMPI_Imrecv(OMPI_F2C_BOTTOM(buf), c_count,
```
[the remark this answers is not in the document]
Comment (ggouaillardet) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/imrecv_cdesc.c]: these files were copied and then edited from `ompi/mpi/fortran/mpif-h` and this is where the error came from. I fixed the `mpif-h` bindings in https://github.com/open-mpi/ompi/pull/6812 and will update this PR accordingly.

**Comment by ibm-ompi** [comment]
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

**Comment by jjhursey** [comment]
bot:ibm:xl:retest

File: ompi/mpi/fortran/use-mpi-f08/cdesc/bsend_init_cdesc.c [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    void *buf = x->base_addr;
+    int c_count =  OMPI_FINT_2_INT(*count);
+
+    c_comm = PMPI_Comm_f2c (*comm);
+
+    c_datatype = c_type;
+    if (x->rank != 0 && !CFI_is_contiguous(x)) {
+        c_ierr = ompi_cdesc_create_datatype(x, c_count, c_type, &c_datatype);
+        if (MPI_SUCCESS != c_ierr) {
+            if (NULL != ierr) *ierr = OMPI_INT_2_FINT(c_ierr);
```
[the remark this answers is not in the document]
Comment (ggouaillardet) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/bsend_init_cdesc.c]: about that ...

there is a lot of copy/paste from `mpifh` (and there are no `{}` here)

but the real question is do we really need to test `if (NULL != ierr)` ?
`ierr` cannot be `NULL` if invoked from `mpif-h` or `usempi`.
If the Fortran parameter is omitted in `usempif08`, then we still invoke this subroutine with an address on the stack.

[the remark this answers is not in the document]
Comment (jsquyres) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/bsend_init_cdesc.c]: Good point.  I think you're right -- we might as well remove all the `if (NULL != ierr)` from this layer because the upper layer guarantees to call with a valid (non-NULL) `ierr`.

File: ompi/mpi/fortran/use-mpi-f08/cdesc/bsend_init_cdesc.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (jsquyres) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/bsend_init_cdesc.c]: If it has to be a macro, that's probably still better than all the duplicated code.

...but just so that the question is asked 😉 -- is it possible to restructure a little bit to not require a `return` / allow it to be a proper function?

File: ompi/mpi/fortran/use-mpi-f08/cdesc/pack_cdesc.c (hunk 10 above)
Comment (jsquyres) [reply ompi/mpi/fortran/use-mpi-f08/cdesc/pack_cdesc.c]: I do not think that there is a restriction on the outbuf to be contiguous -- i.e., it can be a descriptor.  More comments below in the main thread...

**Comment by jsquyres** [comment]
> Honestly, I am not sure I see what the outcome of `mpi_send(a(1:2,1:2), 3, MPI_INT, ...)` is expected to be ? 

The behavior is described on MPI-3.1 p634.  Line 18 starts a particularly interesting section:

> All nonblocking MPI functions (e.g., MPI_ISEND, MPI_PUT, MPI_FILE_WRITE_ALL_BEGIN) behave as if the user-specified elements of choice buffers are copied to a contiguous scratch buffer in the MPI runtime environment. All datatype descriptions (in the example above, “3, MPI_REAL”) read and store data from and to this virtual contiguous scratch buffer. Displacements in MPI derived datatypes are relative to the beginning of this virtual contiguous scratch buffer.

I.e., it is useful to couch all of these conversations in the context of this "virtual contiguous scratch buffer".

Specifically: for all of Gille's shape issues: I unfortunately think they all must be supported.  I do not find any text in MPI-3.1 that says that they can be unsupported by the implementation.  They may not be *performant* (e.g., if someone does a pack from a [non-contig] descriptor to a [non-contig] descriptor -- the implementation, for simplicity, pack from the [non-contig] descriptor to a contiguous buffer and then from that contiguous buffer to the [non-contig] descriptor), but we're talking correctness here, not performance.  I think there's other places in the MPI spec that allow users to do non-performant things, too.

-----

Also, I noticed this morning that per MPI-3.1 table 17.1, we have to change the name of the back-end Fortran symbols if descriptors are supported: they need to end in `_f08ts` (vs. plain `_f08`).  

The spec unfortunately says that we have to support descriptors in all of mpif.h, the mpi module, and the mpi_f08 module.  I don't think we can support it in *just* the mpi_f08 module.  😦 

Finally: per compilers that support `type(*), dimension(..)` but do _not_ support ignore TKR, I guess the question is: do we care about those compilers?  (I do not know which one @ggouaillardet has in mind) . And if so, how hard is it to remove "ignore TKR" from our implementation?  But more specifically: if we have to support `type(*), dimension(..)` in all of mpif.h/mpi/mpi_f08, does ignore TKR matter?

**Comment by ggouaillardet** [comment]
Thanks Jeff,

 - I will do the rename from `_f08` to `_f08ts`
 - I do have a compiler in mind, but I cannot discuss it publically. Since moving to descriptors break the ABI, I already moved to using descriptors everywhere so we do not need to change the ABI again in the future. Some operations simply return `MPI_ERR_INTERN` if the buffer shape is not supported.
 - I will add support for all shapes (fallback to a contiguous copy if we cannot create the right derived datatype). this will require a few logic to be added into the non blocking operations (to free or copy-back the non-contiguous buffer when the operation completes) and causes a few headaches to correctly support persistent operations (e.g. perform a copy in at `MPI_Start()` time).
 - _we have to support descriptors in all of mpif.h, the mpi module, and the mpi_f08 module_ I am a bit puzzled with that one. since `mpif.h` has no prototypes, how can we tell the compiler when an argument should be passed via a descriptor (e.g. a buffer) or its address (e.g. an `INTEGER`) ? I do not have a similar objection with `use mpi`.

FWIW, I have made some progresses with this PR, and my dev branch is at https://github.com/ggouaillardet/ompi/tree/dev/f08_cdesc
(I am also fixing `mpif-h` bindings that do not correctly ignore some parameters, and I created a bunch of tests at https://github.com/ggouaillardet/ompi-tests/tree/topic/ibm_collective_fortran_ignored_params)

**Comment by jsquyres** [comment]
> * I do have a compiler in mind, but I cannot discuss it publically. Since moving to descriptors break the ABI, I already moved to using descriptors everywhere so we do not need to change the ABI again in the future. Some operations simply return `MPI_ERR_INTERN` if the buffer shape is not supported.

Is there a way to return something more descriptive than `MPI_ERR_INTERN` -- i.e., something that indicates that the type is simply not supported by Open MPI?  (vs., for example, an error in the application or an error in Open MPI)  I ask simply to help us save time in the future if/when someone reports this "error", we can clearly tell "oh, ya, that's because it's not supported", as opposed to asking for a small reproducer, and/or looking for an error in Open MPI's code base.

> * I will add support for all shapes (fallback to a contiguous copy if we cannot create the right derived datatype). this will require a few logic to be added into the non blocking operations (to free or copy-back the non-contiguous buffer when the operation completes) and causes a few headaches to correctly support persistent operations (e.g. perform a copy in at `MPI_Start()` time).

I think that when we wrote this text for MPI-3.0, we were envisioning one possible implementation where a "union" datatype was created: i.e., take the "union" of the shape of the data in memory and the supplied MPI datatype, and create a new datatype representing the resulting shape.  Then simply pass that datatype through the MPI engine.  Hypothetically, one would therefore not need to modify the internal MPI engine at all -- because the result of this operation is just another MPI datatype, like any other MPI datatype.

Is that possible here?

> * _we have to support descriptors in all of mpif.h, the mpi module, and the mpi_f08 module_ I am a bit puzzled with that one. since `mpif.h` has no prototypes, how can we tell the compiler when an argument should be passed via a descriptor (e.g. a buffer) or its address (e.g. an `INTEGER`) ? I do not have a similar objection with `use mpi`.

It is _optional_ for `mpif.h` to have prototypes (see MPI-3.1 p611:33-34).  I'm not aware of any MPI implementation that has them, but the spec allows for it.

Re-reading MPI-3.1 a little better than I did yesterday, I notice this (starting on p613:47):

> To set MPI_SUBARRAYS_SUPPORTED to .TRUE. within a Fortran support method, it is required that all non-blocking and split-collective routines with buffer arguments are implemented according to 1B and 2B, i.e., with MPI_Xxxx_f08ts in the mpi_f08 module, and with MPI_XXXX_FTS in the mpi module and the mpif.h include file.
> The mpi and mpi_f08 modules and the mpif.h include file will each correspond to exactly one implementation scheme from Table 17.1. However, the MPI library may contain multiple implementation schemes from Table 17.1.

This section -- and sections 17.1.2/3/4 -- says to me that each of `mpif.h`, the `mpi` module, and `mpi_f08` modules can independently set `MPI_SUBARRAYS_SUPPORTED` to different values.

I guess that makes sense for exactly the case you are asking about: we can have `mpi_f08` support `TYPE(*),DIMENSION(..)`, but `mpi` and `mpif.h` do not.

...but how do we implement that?  The name `MPI_SUBARRAYS_SUPPORTED` has a single back-end symbol: how can we have it have different values depending on whether you use `mpif.h`/`mpi`/`mpi_f08`?

Regardless, this opens a whole new can of worms:

* Do we continue to have the non-descriptor-enabled symbols?  E.g., do we have both the descriptor and non-descriptor symbols in `libomi_mpif.so`?
* Do we have a `configure` option (or compile-type option?) to select between `SUBARRAYS_SUPPORTED==.TRUE.` or `.FALSE.`?
* How many MPI tools supports the descriptor-enabled symbols?
* Is it going to cause an armageddon for user apps if we start putting prototypes in `mpif.h`?

😲

**Comment by ggouaillardet** [comment]
would `MPI_ERR_UNSUPPORTED_DATAREP` or `MPI_ERR_UNSUPPORTED_OPERATION` be a better fit compared to `MPI_ERR_INTERN` ? If not, would it be MPI compliant to define a new error code? (or can we simply display an ad-hoc error message and return with an existing error code?)



I do not fully understand what you mean by "union" datatype.
Here is a corner-case though

```
   INTEGER :: buf(0:2,1:3)
   CALL MPI_GATHER(..., recvbuf=buf(1:2, 1:3), recvcount=6, recvtype=MPI_INTEGER, ...)
```

and run this on a communicator with two MPI ranks.
If I get it right, that would require a `MPI_Gatherw` like operation that is not even part of the standard.

Note I chose to have `MPI_Win_create` errors with `MPI_ERR_BUFFER` if a user tries to use a subarray as the `base` argument.



I'd rather start with `ts` only for the `mpi_f08` bindings.

> ...but how do we implement that? The name MPI_SUBARRAYS_SUPPORTED has a single back-end symbol

the definition is currently in a single place (e.g. `mpif-config.h`) and this file is currently included within `module mpi` and `module mpi_f08`.
I made a simple proof-of-concept (e.g. move the definition to an other file, and have only `mpi_f08` include its own version. Bottom line, `MPI_SUBARRAYS_SUPPORTED` is `.FALSE.` for `mpif-h` and `use-mpi` but `.TRUE.` for `use-mpi-f08`.



We do need a configure option to manually disable `ts` bindings :
I found a blocker in gcc 9 that makes the bindings unusable ... see my report at https://gcc.gnu.org/ml/fortran/2019-08/msg00104.html



We should decide asap if we want to have both `MPI_xyz_f08` and `MPI_xyz_f08ts` symbols in the library. I was planning to move `xyz_f08.F90` into `xyz_f08.F90.in` to have the stuff generated at configure time and make it easier to troubleshoot. Obviously, that won't work if we want to have both symbols in the library.

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/01f8be4f3c3350eff5399cfba60efac9

**Comment by jsquyres** [comment]
This issue kinda fell off the table for a few months.  ☹️ 

@ggouaillardet I don't see a reply to your issue on the GCC mailing list.  Did that get resolved?

Given the lateness in the game here, I'm kinda guessing that this is going to miss the boat for Open MPI v5.0.0.

**Comment by gpaulsen** [comment]
Discussed on call today.  If this can make it in by v5.0.x branch-date (5/14/2020), great, but probably next opportunity will be v6.0.0 (currently unscheduled).

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/09792ae4918db9664b9dc6145ebfe8e2

**Comment by awehrfritz** [comment]
It's been a while since there has been any activity on this PR. Is there any interest in getting this feature set merged at some point?

**Comment by hppritcha** [comment]
this work has been superseded by PR #12226 which was merged in to main a while ago now.  closing this PR.
